File: backend/src/modules/fraud/fraud.service.js
// ============================================================
// ARTH Backend — Fraud Detection Service
// Analyzes transaction behavior and calls ML for fraud risk.
// ============================================================

const mlClient = require('../../integrations/mlClient');
const transactionService = require('../transaction/transaction.service');
const logger = require('../../utils/logger');

/**
 * Detect fraud risk for a transaction or user behavior.
 */
const detect = async (userId, transactionData = {}) => {
  const summary = await transactionService.getTransactionSummary(userId);

  // Calculate behavioral features for fraud detection
  const avgAmount = summary.transactionCount > 0
    ? (summary.totalIncome + summary.totalExpenses) / summary.transactionCount
    : 0;

  const payload = {
    // Transaction-level features
    amount: transactionData.amount || 0,
    category: transactionData.category || 'unknown',
    location: transactionData.location || 'unknown',
    device: transactionData.device || 'unknown',
    hour_of_day: transactionData.hour_of_day || new Date().getHours(),

    // User behavioral features
    avg_transaction_amount: avgAmount,
    transaction_count: summary.transactionCount,
    total_income: summary.totalIncome,
    total_expenses: summary.totalExpenses,

    // Anomaly indicators
    amount_deviation: avgAmount > 0
      ? Math.abs((transactionData.amount || 0) - avgAmount) / avgAmount
      : 0,
    is_new_category: transactionData.category
      ? !summary.categoryBreakdown[transactionData.category]
      : false,
  };

  logger.info(`Fraud detection request for user ${userId}`, {
    amount: payload.amount,
    category: payload.category,
  });

  const result = await mlClient.detectFraud(payload);

  return {
    fraudAnalysis: result,
    transactionDetails: {
      amount: payload.amount,
      category: payload.category,
      location: payload.location,
    },
    behavioralContext: {
      avgTransactionAmount: avgAmount,
      totalTransactions: summary.transactionCount,
    },
  };
};

module.exports = { detect };

File: ml-service/app/services/fraud_service.py
# ============================================================
# ARTH ML Service — Fraud Service
# ============================================================

from app.models.fraud_model import fraud_model
from app.schemas.fraud import FraudRequest
import logging

logger = logging.getLogger("arth.ml.fraud.service")


def detect(request: FraudRequest) -> dict:
    """Detect fraud risk for a transaction."""
    try:
        logger.info(f"Running fraud detection: amount={request.amount}, category={request.category}")

        features = {
            "amount": request.amount,
            "hour_of_day": request.hour_of_day,
            "avg_transaction_amount": request.avg_transaction_amount,
            "amount_deviation": request.amount_deviation,
            "transaction_count": request.transaction_count,
            "is_new_category": request.is_new_category,
        }

        result = fraud_model.detect(features)
        logger.info(f"Fraud detection result: prob={result['fraud_probability']}, level={result['risk_level']}")
        return result

    except Exception as e:
        logger.error(f"Fraud detection failed: {str(e)}")
        raise
